codegen-core/src/main/kotlin/software/amazon/smithy/rust/codegen/core/smithy/generators/protocol/ProtocolTestGenerator.kt

```diff
@@ -0,0 +1,325 @@
+import software.amazon.smithy.protocoltests.traits.HttpMalformedRequestTestsTrait
+import software.amazon.smithy.protocoltests.traits.HttpRequestTestCase
+import software.amazon.smithy.protocoltests.traits.HttpRequestTestsTrait
+import software.amazon.smithy.protocoltests.traits.HttpResponseTestCase
+import software.amazon.smithy.protocoltests.traits.HttpResponseTestsTrait
+import software.amazon.smithy.rust.codegen.core.rustlang.Attribute
+import software.amazon.smithy.rust.codegen.core.rustlang.Attribute.Companion.allow
+import software.amazon.smithy.rust.codegen.core.rustlang.Attribute.Companion.shouldPanic
+import software.amazon.smithy.rust.codegen.core.rustlang.CargoDependency
+import software.amazon.smithy.rust.codegen.core.rustlang.RustModule
+import software.amazon.smithy.rust.codegen.core.rustlang.RustWriter
+import software.amazon.smithy.rust.codegen.core.rustlang.Writable
+import software.amazon.smithy.rust.codegen.core.rustlang.docs
+import software.amazon.smithy.rust.codegen.core.rustlang.rust
+import software.amazon.smithy.rust.codegen.core.rustlang.rustBlock
+import software.amazon.smithy.rust.codegen.core.rustlang.rustTemplate
+import software.amazon.smithy.rust.codegen.core.rustlang.withBlock
+import software.amazon.smithy.rust.codegen.core.smithy.CodegenContext
+import software.amazon.smithy.rust.codegen.core.smithy.RuntimeType
+import software.amazon.smithy.rust.codegen.core.testutil.testDependenciesOnly
+import software.amazon.smithy.rust.codegen.core.util.dq
+import software.amazon.smithy.rust.codegen.core.util.getTrait
+import software.amazon.smithy.rust.codegen.core.util.orNull
+import software.amazon.smithy.rust.codegen.core.util.outputShape
+import software.amazon.smithy.rust.codegen.core.util.toSnakeCase
+
+/**
+ * Common interface to generate protocol tests for a given [operationShape].
+ */
+interface ProtocolTestGenerator {
+    val codegenContext: CodegenContext
+    val protocolSupport: ProtocolSupport
+    val operationShape: OperationShape
+
+    /**
+     * We expect these tests to fail due to shortcomings in our implementation.
+     * They will _fail_ if they pass, so we will discover and remove them if we fix them by accident.
+     **/
+    val expectFail: Set<FailingTest>
+
+    /** Only generate these tests; useful to temporarily set and shorten development cycles */
+    val runOnly: Set<String>
+
+    /**
+     * These tests are not even attempted to be generated, either because they will not compile
+     * or because they are flaky.
+     */
+    val disabledTests: Set<String>
+
+    /** The Rust module in which we should generate the protocol tests for [operationShape]. */
+    fun protocolTestsModule(): RustModule.LeafModule {
+        val operationName = codegenContext.symbolProvider.toSymbol(operationShape).name
+        val testModuleName = "${operationName.toSnakeCase()}_test"
+        val additionalAttributes =
+            listOf(Attribute(allow("unreachable_code", "unused_variables")))
```

Comment:
Because we generate code like:

```rust
async move {
    let result = { panic!("request should have been rejected, but we accepted it; we parsed operation input `{:?}`", &input) as crate::output::MalformedRequestBodyOutput };
    sender.send(()).await.expect("receiver dropped early");
    result
}
```

And the compiler flags the `as` cast as being unreachable, but we need it for things to type-check.